src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,512 +158,931 @@ enum Spells
         {
-            Initialize();
+            return GetTrialOfChampionAI<npc_fountain_of_lightAI>(creature);
         }
+};
 
-        void UpdateAI(uint32 uiDiff) override
+class npc_memory : public CreatureScript
+{
+    public:
+        npc_memory() : CreatureScript("npc_memory") { }
+
+        struct npc_memoryAI : public ScriptedAI
         {
-            if (!UpdateVictim())
-                return;
+            npc_memoryAI(Creature* creature) : ScriptedAI(creature)
+            {
+                Initialize();
+				instance = creature->GetInstanceScript();
+                me->SetReactState(REACT_PASSIVE);
+            }
+			
+			InstanceScript* instance;
+			
+            void Initialize()
+            {
+                DoCastAOE(SPELL_SHADOWFORM);
+                DoCastAOE(SPELL_SPAWN_VISUAL);
+                _events.ScheduleEvent(EVENT_ENTER_AGGRESSIVE, 3 * IN_MILLISECONDS);
+            }
 
-            if (uiOldWoundsTimer <= uiDiff)
+            void Reset() override
             {
-                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0))
-                {
-                    if (target->IsAlive())
-                        DoCast(target, SPELL_OLD_WOUNDS);
-                }
-                uiOldWoundsTimer = 12000;
-            }else uiOldWoundsTimer -= uiDiff;
+                _events.Reset();
+                Initialize();
+            }
+
+            void JustDied(Unit* /*killer*/) override
+            {
+                /*if (Unit* summoner = me->ToTempSummon()->GetSummoner())
+                    summoner->GetAI()->DoAction(ACTION_MEMORY_DEATH);*/
+                /*if (Creature* pale = instance->GetCreature(NPC_PALETRESS)) {
+                    pale->AI()->DoAction(ACTION_MEMORY_DEATH);
+                }*/
+				if (Unit* summoner = me->ToTempSummon()->GetSummoner()) {
+                    if (Creature* pale = summoner->ToCreature()) {
+						pale->AI()->DoAction(ACTION_MEMORY_DEATH);
+					}
+				}
+				
+                me->DespawnOrUnsummon();
+            }
 
-            if (uiWakingNightmare <= uiDiff)
+            void JustEngagedWith(Unit* /*who*/) override
             {
-                DoCast(me, SPELL_WAKING_NIGHTMARE);
-                uiWakingNightmare = 7000;
-            }else uiWakingNightmare -= uiDiff;
+                DoZoneInCombat();
+            }
 
-            if (uiShadowPastTimer <= uiDiff)
+            void UpdateAI(uint32 diff) override
             {
-                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 1))
+                _events.Update(diff);
+
+                if (me->HasUnitState(UNIT_STATE_CASTING))
+                    return;
+
+                while (uint32 eventId = _events.ExecuteEvent())
                 {
-                    if (target->IsAlive())
-                        DoCast(target, SPELL_SHADOWS_PAST);
+                    switch (eventId)
+                    {
+                        case EVENT_ENTER_AGGRESSIVE:
```

Comment:
I don't really like this event, because it requires `UpdateVictim` to be after the eventmap update. Can we move it to a timer on Palatress' AI instead?

Reply:
@Treeston What would be the right way to do this? using the same event already on the Paletress AI or adding another event on the AI? if so would this imply having 2 different whiles on the UpdateAI of the Paletress? or i got all wrong? maybe you remember a script were i can look for reference?

Reply:
Huh? Just add a second event to Palatress' `enum`, and a `case` for it on Palatress' AI, and schedule that event however many seconds after the initial summon.

Reply:
I think i got it now. thank you again!